src/parser/DASHTree.cpp

```diff
@@ -1775,19 +1775,54 @@ void DASHTree::RefreshLiveSegments()
 
       for (size_t index{0}; index < updateTree->periods_.size(); index++)
       {
-        if (index == periods_.size()) // Exceeded periods
-          break;
         auto updPeriod = updateTree->periods_[index];
         if (!updPeriod)
           continue;
 
+        Period *period{nullptr};
+        // find matching period based on ID
+        auto itPd = std::find_if(periods_.begin(), periods_.end(),
+                                 [&updPeriod](const Period* item)
+                                 {
+                                   return !item->id_.empty() && item->id_ == updPeriod->id_;
+                                 });
+        if (itPd == periods_.end())
+        {
+          // not found, try matching period based on start
+          itPd = std::find_if(periods_.begin(), periods_.end(),
+                              [&updPeriod](const Period* item)
+                              {
+                                return item->start_ && item->start_ == updPeriod->start_;
+                              });
+
+        }
+        // found!
+        if (itPd != periods_.end())
+          period = *itPd;
+        if (!period && updPeriod->id_.empty() && updPeriod->start_ == 0)
+        {
+          // not found, fallback match based on position
+          if (index < periods_.size())
+            period = periods_[index];
+        }
+        // new period, insert it
+        if (!period)
+        {
+          LOG::LogF(LOGDEBUG, "Inserting new Period (id=%s, start=%ld)", updPeriod->id_.c_str(), updPeriod->start_);
+          std::vector<Period*> tmp = periods_;
```

Comment:
im not able to understand why you need to copy the vector of pointers
can you explain?

Reply:
May not be required, I will test without it.

Reply:
The reason is because periods_ is used in a thread-unsafe manner (see comment at the top of the function).

So other threads from session/main may be iterating the vector, and if I alter the vector in-place then those iterators are invalidated and will cause exceptions.

So it is simpler to make a copy and then replace for now, until thread safety is added.

Reply:
yeah the comment on top is a my old comment as todo to check,
this method is called behind a `GetTreeMutex()` in this case the "tree" variables should be safe,
but, there may be old oversights, now long time passed and i dont remember anymore exaclty what was needed to check,
looking at PR #1121 can give an idea where is needed add missing mutex's

now i am not asking you to solve the problem on this PR
but at this point its better add a todo comment above the line
`std::vector<Period*> tmp`
similar to:
`//! @todo: this temporary variable can be safely removed when this method will be thread safe`
in this way will not be forgotten the reason of this when we have to cleanup